js/client-mainmenu.js

```diff
@@ -114,7 +125,68 @@
 				);
 			}
 		},
+		createLoginFormObserver:function (self) {
+			const config = {attributes: true, childList: true, subtree: true};
+			const observer = new MutationObserver(function () {
+				if (window.app.user.get("named") && window.Storage.prefs("page") == "challenge") {
+					window.Storage.prefs("page", "testclient");
+					if ($('.challenge-container').length == 0) {
+						//user log in event trigerd when user log in and make user pick a team
+						$('body').on("login", function () {
+							self.onUserLogIn(self);
+						});
+						$('body').trigger("login");
+					}
+				}
+			});
+			observer.observe(document.activeElement, config);
+		},
+		onUserLogIn:function (self) {
+			let userId = window.Storage.prefs("challenger");
+			let format = window.Storage.prefs("format");
+			app.rooms[''].challenge(userId, format, (format.includes("random") ? '' : format));
+			$('.select.formatselect').prop("disabled",true);
+		},
+		generateLink:function (userId, format) {
+			if (!$('.link').length) {
+				let link = window.location.protocol+"//" + window.Config.routes.client + "?challenge-" + userId + "-" + format;
+				$('.link-generatore').append("<p class=link>" + link + "</p>");
+				$('#generate-link').html("generate link");
+			} else {
+				let link = window.location.protocol+"//" + window.Config.routes.client + "?challenge-" + userId + "-" + format;
+				$('.link').html("<p class=link>" + link + "</p>");
+				$('#generate-link').html("generate link");
+			}
+		},
+		generateChallengeForm:function (self) {
+			var buf = '<div class="challenge"><form class="battleform"><p>Challenge link generatore</p>';
+			buf += '<p><label class="label">Format:</label>' + self.renderFormats("gen8randombattle") + '</p>';
+			buf += '<button type="button" id=generate-link>generate link</button>';
+			buf += '<button type="button" id=close-challenge>close</button></form></div>';
```

Comment:
```suggestion
			buf += '<button type="button" id=close-challenge>close</button></form></div>';
```
Element ids are bad practice in this project - you should be using the element's name and then checking which element it is in jQuery event listeners that are set on it.